docs/_pages/grammar.md

```diff
@@ -68,12 +68,17 @@ SimpleType =
 SingletonType = STRING | 'true' | 'false'
 
 Type =
-    SimpleType ['?'] |
+    Type ['?'] |
     Type ['|' Type] |
-    Type ['&' Type]
+    Type ['&' Type] |
+    '(' Type ')'
```

Comment:
I believe this is correct, since it previously did not allow the following:
- `string??`
- `(string)`
- `(string | number) & boolean`

Did some quick checking and it seems alright, but not sure if its missing any edge cases.

Also, the grammar does not encode the fact that `string | number & boolean` is not allowed (one of the union or intersection needs to be surrounded by parentheses). Should it do this?

Reply:
This new definition lost the reference to `SimpleType` completely.
And `(Type)` handling should go into `SimpleType`.

After that, specifying the limitation should be easier, along the lines of
```
UnionSuffix = {'?'} ['|' SimpleType]
IntersectionSuffix = ['&' SimpleType]
Type = SimpleType {UnionSuffix | IntersectionSuffix}
```

Reply:
I don't see how my definition allows that, `UnionSuffix` doesn't continue into `Type`, so it can no longer reach `IntersectionSuffix` (and the same holds for `IntersectionSuffix`).
Maybe I'm wrong about the syntax, but I thought `A (B | C) == (A B) | (A C)`.

Reply:
That's correct for `A (B | C)`, but here it is `A {B | C}` (repetition), so `A B B C C B` would be allowed I believe

If we switch the repetitions around it might work? Like
```
UnionSuffix = {'?'} {'|' SimpleType}      # changed from ['|' SimpleType], and same below
IntersectionSuffix = {'&' SimpleType}
Type = SimpleType (UnionSuffix | IntersectionSuffix)
```